### PR TITLE
Feature/alias

### PR DESCRIPTION
**Pull request by drewbanin**
Reopened from https://github.com/fishtown-analytics/dbt/pull/651, https://github.com/fishtown-analytics/dbt/pull/771

Description (and original code courtesy of @abelsonlive). Thanks also to @jon-rtr for helping bring this branch up-to-date with development.

-------

# WHAT

This P.R. implements model aliasing RE: https://github.com/fishtown-analytics/dbt/issues/637.

# WHY 

To enable more control over how `dbt` materializes tables.

# HOW 

This P.R. adds an `alias` parameter to the model `config`. If not present, this parameter will default to the `name` of the model (AKA the parsed file path of the model). This implementation is preferable because it maintains the use of the filename in `ref` calls, thereby enabling `dbt` users to have distinct model names, but non-distinct table names. 

...

Model aliasing is implemented for seed and model resources. An `alias` can be specified in `dbt_project.yml` or in a `{{ config(alias=...) }}` block. The `name` attribute of a Relation object will reflect the alias, so `{{ this.name }}` will return a model alias if one is present.

### CODE REVIEW

**Comment by abelsonlive** [comment]
Wooooot!

**Comment by cmcarthur** [comment]
long-run we should replace `node['alias']` with `model.alias` but whatever.

unrelated: your build is failing.

**Comment by drewbanin** [comment]
@cmcarthur ok, I ripped out `get_custom_alias`. I think the better approach is to jinja-compile the `alias` field, but we should likely do that for _many_ dbt fields, so I think we can handle that in a different PR. I was trying to achieve a way to dynamically change the name of the resulting table for eg. date partitioning on BQ. I definitely agree that it was not the right approach here

**Comment by drewbanin** [comment]
these tests are failing with a _very_ upsetting:
```
On expect_value_foo_alias_tablename__foo: 

select count(*)
from "test1530625544_aliases_026"."foo"
where tablename != 'foo'

dbt: DEBUG: Postgres error: failed to find conversion function from unknown to text
```

This works locally for me, so I think it must be a postgres version thing. Just going to keep adding typecasts until it works /shrug

File: test/integration/026_aliases_test/macros/cast.sql [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  {{ adapter_macro('test.string_literal', s) }}
+{%- endmacro %}
+
+{% macro default__string_literal(s) %}
+    '{{ s }}'::text
+{% endmacro %}
+
+{% macro bigquery__string_literal(s) %}
+    cast('{{ s }}' as string)
+{% endmacro %}
```
Comment (cmcarthur) [inline test/integration/026_aliases_test/macros/cast.sql]: random: let's move this macro https://github.com/fishtown-analytics/dbt-utils/blob/master/macros/cross_db_utils/safe_cast.sql from dbt-utils to dbt proper.

Comment (drewbanin) [reply test/integration/026_aliases_test/macros/cast.sql]: yeah, agree, that would be super useful. Do you think that should happen in this PR or a different one?

File: test/integration/026_aliases_test/test_aliases.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    @property
+    def models(self):
+        return "test/integration/026_aliases_test/models-dupe-custom-schema"
+
+    @property
+    def project_config(self):
+        return {
+            "macro-paths": ['test/integration/026_aliases_test/macros'],
+        }
```
Comment (cmcarthur) [inline test/integration/026_aliases_test/test_aliases.py]: i really like this as a test case for this feature

**Comment by cmcarthur** [comment]
@drewbanin this looks great. i left a couple random thoughts here but you can ship this.

**Pull request merged**